File: Task2/lru.py
"""
ST5004CEM Operating Systems and Security
Task 2 - Least Recently Used (LRU) Page Replacement

Author: Trisha Rajbanshi
"""

def lru_page_replacement(reference_string, frames):

    memory = []
    page_faults = 0

    print("=" * 60)
    print("LEAST RECENTLY USED (LRU) PAGE REPLACEMENT")
    print("=" * 60)

    for page in reference_string:

        # Page Hit
        if page in memory:
            memory.remove(page)
            memory.append(page)
            print(f"Page {page} -> HIT")

        else:
            page_faults += 1

            if len(memory) == frames:
                removed = memory.pop(0)
                print(f"Removed Page: {removed}")

            memory.append(page)
            print(f"Inserted Page: {page}")

        print("Frames:", memory)
        print("-" * 40)

    print("=" * 60)
    print(f"Total Page Faults : {page_faults}")
    print("=" * 60)


def main():

    reference_string = [7, 0, 1, 2, 0, 3, 0, 4, 2, 3, 0, 3, 2]

    frames = 3

    lru_page_replacement(reference_string, frames)


if __name__ == "__main__":
    main()

File: Task2/lru.c
#include <stdio.h>

int main()
{
    int pages[50], frames[10], recent[10];
    int n, f;
    int pageFaults = 0;
    int time = 0;

    printf("===== LRU Page Replacement =====\n\n");

    printf("Enter number of pages: ");
    scanf("%d", &n);

    printf("Enter number of frames: ");
    scanf("%d", &f);

    printf("Enter page reference string:\n");

    for(int i = 0; i < n; i++)
        scanf("%d", &pages[i]);

    for(int i = 0; i < f; i++)
    {
        frames[i] = -1;
        recent[i] = -1;
    }

    printf("\nPage\tFrames\n");

    for(int i = 0; i < n; i++)
    {
        int found = 0;

        for(int j = 0; j < f; j++)
        {
            if(frames[j] == pages[i])
            {
                found = 1;
                recent[j] = ++time;
                break;
            }
        }

        if(!found)
        {
            int position = 0;

            for(int j = 1; j < f; j++)
            {
                if(recent[j] < recent[position])
                    position = j;
            }

            frames[position] = pages[i];
            recent[position] = ++time;
            pageFaults++;
        }

        printf("%d\t", pages[i]);

        for(int j = 0; j < f; j++)
        {
            if(frames[j] == -1)
                printf("- ");
            else
                printf("%d ", frames[j]);
        }

        printf("\n");
    }

    printf("\nTotal Page Faults : %d\n", pageFaults);

    return 0;
}
